File: src/db/schema.ts
export const SCHEMA_STATEMENTS = [
  `CREATE TABLE IF NOT EXISTS embedding_models (
    id TEXT PRIMARY KEY,
    name TEXT NOT NULL,
    provider TEXT NOT NULL,
    dimensions INTEGER NOT NULL,
    is_active INTEGER DEFAULT 0,
    created_at INTEGER NOT NULL DEFAULT (unixepoch() * 1000)
)`,
  `CREATE TABLE IF NOT EXISTS projects (
    id TEXT PRIMARY KEY,
    path TEXT NOT NULL UNIQUE,
    name TEXT,
    settings_json TEXT,
    created_at INTEGER NOT NULL DEFAULT (unixepoch() * 1000),
    updated_at INTEGER NOT NULL DEFAULT (unixepoch() * 1000)
)`,
  `CREATE TABLE IF NOT EXISTS sessions (
    id TEXT PRIMARY KEY,
    project_id TEXT NOT NULL,
    started_at INTEGER NOT NULL,
    ended_at INTEGER,
    summary TEXT,
    user_prompt TEXT,
    context_json TEXT,
    FOREIGN KEY (project_id) REFERENCES projects(id)
)`,
  `CREATE TABLE IF NOT EXISTS memories (
    id TEXT PRIMARY KEY,
    project_id TEXT NOT NULL,
    content TEXT NOT NULL,
    summary TEXT,
    content_hash TEXT,
    sector TEXT NOT NULL,
    tier TEXT DEFAULT 'project', -- 'session' | 'project'
    importance REAL DEFAULT 0.5,
    categories_json TEXT,
    simhash TEXT,
    salience REAL DEFAULT 1.0,
    access_count INTEGER DEFAULT 0,
    created_at INTEGER NOT NULL,
    updated_at INTEGER NOT NULL,
    last_accessed INTEGER NOT NULL,
    valid_from INTEGER,
    valid_until INTEGER,
    is_deleted INTEGER DEFAULT 0,
    deleted_at INTEGER,
    embedding_model_id TEXT,
    tags_json TEXT,
    concepts_json TEXT,
    files_json TEXT,
    FOREIGN KEY (project_id) REFERENCES projects(id),
    FOREIGN KEY (embedding_model_id) REFERENCES embedding_models(id)
)`,
  `CREATE TABLE IF NOT EXISTS session_memories (
    session_id TEXT NOT NULL,
    memory_id TEXT NOT NULL,
    created_at INTEGER NOT NULL,
    usage_type TEXT NOT NULL,
    PRIMARY KEY (session_id, memory_id, created_at),
    FOREIGN KEY (session_id) REFERENCES sessions(id),
    FOREIGN KEY (memory_id) REFERENCES memories(id)
)`,
  `CREATE TABLE IF NOT EXISTS entities (
    id TEXT PRIMARY KEY,
    name TEXT NOT NULL,
    entity_type TEXT NOT NULL,
    summary TEXT,
    embedding_model_id TEXT,
    created_at INTEGER NOT NULL,
    updated_at INTEGER NOT NULL,
    FOREIGN KEY (embedding_model_id) REFERENCES embedding_models(id)
)`,
  `CREATE TABLE IF NOT EXISTS memory_entities (
    memory_id TEXT NOT NULL,
    entity_id TEXT NOT NULL,
    role TEXT NOT NULL,
    created_at INTEGER NOT NULL,
    PRIMARY KEY (memory_id, entity_id),
    FOREIGN KEY (memory_id) REFERENCES memories(id),
    FOREIGN KEY (entity_id) REFERENCES entities(id)
)`,
  `CREATE TABLE IF NOT EXISTS memory_relationships (
    id TEXT PRIMARY KEY,
    source_memory_id TEXT NOT NULL,
    target_memory_id TEXT NOT NULL,
    relationship_type TEXT NOT NULL,
    created_at INTEGER NOT NULL,
    valid_from INTEGER NOT NULL,
    valid_until INTEGER,
    confidence REAL DEFAULT 1.0,
    extracted_by TEXT NOT NULL,
    FOREIGN KEY (source_memory_id) REFERENCES memories(id),
    FOREIGN KEY (target_memory_id) REFERENCES memories(id)
)`,
  `CREATE TABLE IF NOT EXISTS memory_vectors (
    memory_id TEXT PRIMARY KEY,
    model_id TEXT NOT NULL,
    vector F32_BLOB NOT NULL,
    dim INTEGER NOT NULL,
    created_at INTEGER NOT NULL DEFAULT (unixepoch() * 1000),
    FOREIGN KEY (memory_id) REFERENCES memories(id) ON DELETE CASCADE,
    FOREIGN KEY (model_id) REFERENCES embedding_models(id)
)`,
  `CREATE TABLE IF NOT EXISTS documents (
    id TEXT PRIMARY KEY,
    project_id TEXT NOT NULL,
    source_path TEXT,
    source_url TEXT,
    source_type TEXT NOT NULL,
    title TEXT,
    full_content TEXT NOT NULL,
    checksum TEXT,
    created_at INTEGER NOT NULL,
    updated_at INTEGER NOT NULL,
    FOREIGN KEY (project_id) REFERENCES projects(id)
)`,
  `CREATE TABLE IF NOT EXISTS document_chunks (
    id TEXT PRIMARY KEY,
    document_id TEXT NOT NULL,
    chunk_index INTEGER NOT NULL,
    content TEXT NOT NULL,
    start_offset INTEGER,
    end_offset INTEGER,
    tokens_estimate INTEGER,
    FOREIGN KEY (document_id) REFERENCES documents(id) ON DELETE CASCADE
)`,
  `CREATE TABLE IF NOT EXISTS document_vectors (
    chunk_id TEXT PRIMARY KEY,
    model_id TEXT NOT NULL,
    vector F32_BLOB NOT NULL,
    dim INTEGER NOT NULL,
    created_at INTEGER NOT NULL DEFAULT (unixepoch() * 1000),
    FOREIGN KEY (chunk_id) REFERENCES document_chunks(id) ON DELETE CASCADE,
    FOREIGN KEY (model_id) REFERENCES embedding_models(id)
)`,
];

export const FTS_STATEMENTS = [
  `CREATE VIRTUAL TABLE IF NOT EXISTS memories_fts USING fts5(
    content,
    summary,
    concepts_json,
    tags_json,
    content='memories',
    content_rowid='rowid'
)`,
  `CREATE TRIGGER IF NOT EXISTS memories_ai AFTER INSERT ON memories BEGIN
    INSERT INTO memories_fts(rowid, content, summary, concepts_json, tags_json)
    VALUES (NEW.rowid, NEW.content, NEW.summary, NEW.concepts_json, NEW.tags_json);
END`,
  `CREATE TRIGGER IF NOT EXISTS memories_ad AFTER DELETE ON memories BEGIN
    INSERT INTO memories_fts(memories_fts, rowid, content, summary, concepts_json, tags_json)
    VALUES ('delete', OLD.rowid, OLD.content, OLD.summary, OLD.concepts_json, OLD.tags_json);
END`,
  `CREATE TRIGGER IF NOT EXISTS memories_au AFTER UPDATE ON memories BEGIN
    INSERT INTO memories_fts(memories_fts, rowid, content, summary, concepts_json, tags_json)
    VALUES ('delete', OLD.rowid, OLD.content, OLD.summary, OLD.concepts_json, OLD.tags_json);
    INSERT INTO memories_fts(rowid, content, summary, concepts_json, tags_json)
    VALUES (NEW.rowid, NEW.content, NEW.summary, NEW.concepts_json, NEW.tags_json);
END`,
  `CREATE VIRTUAL TABLE IF NOT EXISTS documents_fts USING fts5(
    content,
    content='document_chunks',
    content_rowid='rowid'
)`,
  `CREATE TRIGGER IF NOT EXISTS documents_ai AFTER INSERT ON document_chunks BEGIN
    INSERT INTO documents_fts(rowid, content) VALUES (NEW.rowid, NEW.content);
END`,
  `CREATE TRIGGER IF NOT EXISTS documents_ad AFTER DELETE ON document_chunks BEGIN
    INSERT INTO documents_fts(documents_fts, rowid, content)
    VALUES ('delete', OLD.rowid, OLD.content);
END`,
  `CREATE TRIGGER IF NOT EXISTS documents_au AFTER UPDATE ON document_chunks BEGIN
    INSERT INTO documents_fts(documents_fts, rowid, content)
    VALUES ('delete', OLD.rowid, OLD.content);
    INSERT INTO documents_fts(rowid, content) VALUES (NEW.rowid, NEW.content);
END`,
];

export const EXTRACTION_SCHEMA_STATEMENTS = [
  `CREATE TABLE IF NOT EXISTS extraction_segments (
    id TEXT PRIMARY KEY,
    session_id TEXT NOT NULL,
    project_id TEXT NOT NULL,
    trigger TEXT NOT NULL,
    user_prompts_json TEXT,
    files_read_json TEXT,
    files_modified_json TEXT,
    tool_call_count INTEGER DEFAULT 0,
    memories_extracted INTEGER DEFAULT 0,
    extraction_tokens INTEGER,
    segment_start INTEGER,
    segment_end INTEGER,
    extraction_duration_ms INTEGER,
    created_at INTEGER NOT NULL DEFAULT (unixepoch() * 1000),
    FOREIGN KEY (session_id) REFERENCES sessions(id),
    FOREIGN KEY (project_id) REFERENCES projects(id)
  )`,
  `CREATE TABLE IF NOT EXISTS segment_accumulators (
    session_id TEXT PRIMARY KEY,
    project_id TEXT NOT NULL,
    segment_id TEXT NOT NULL,
    segment_start INTEGER NOT NULL,
    user_prompts_json TEXT DEFAULT '[]',
    files_read_json TEXT DEFAULT '[]',
    files_modified_json TEXT DEFAULT '[]',
    commands_run_json TEXT DEFAULT '[]',
    errors_encountered_json TEXT DEFAULT '[]',
    searches_performed_json TEXT DEFAULT '[]',
    last_assistant_message TEXT,
    tool_call_count INTEGER DEFAULT 0,
    updated_at INTEGER NOT NULL DEFAULT (unixepoch() * 1000),
    FOREIGN KEY (session_id) REFERENCES sessions(id),
    FOREIGN KEY (project_id) REFERENCES projects(id)
  )`,
];

export const INDEX_STATEMENTS = [
  `CREATE INDEX IF NOT EXISTS idx_memories_project ON memories(project_id) WHERE is_deleted = 0`,
  `CREATE INDEX IF NOT EXISTS idx_memories_sector ON memories(sector)`,
  `CREATE INDEX IF NOT EXISTS idx_memories_tier ON memories(tier)`,
  `CREATE INDEX IF NOT EXISTS idx_memories_salience ON memories(salience DESC)`,
  `CREATE INDEX IF NOT EXISTS idx_memories_created ON memories(created_at DESC)`,
  `CREATE INDEX IF NOT EXISTS idx_memories_simhash ON memories(simhash)`,
  `CREATE INDEX IF NOT EXISTS idx_memories_valid ON memories(valid_from, valid_until)`,
  `CREATE INDEX IF NOT EXISTS idx_sessions_project ON sessions(project_id)`,
  `CREATE INDEX IF NOT EXISTS idx_sessions_started ON sessions(started_at DESC)`,
  `CREATE INDEX IF NOT EXISTS idx_session_memories_session ON session_memories(session_id)`,
  `CREATE INDEX IF NOT EXISTS idx_session_memories_memory ON session_memories(memory_id)`,
  `CREATE INDEX IF NOT EXISTS idx_entities_type ON entities(entity_type)`,
  `CREATE INDEX IF NOT EXISTS idx_entities_name ON entities(name)`,
  `CREATE INDEX IF NOT EXISTS idx_memory_entities_memory ON memory_entities(memory_id)`,
  `CREATE INDEX IF NOT EXISTS idx_memory_entities_entity ON memory_entities(entity_id)`,
  `CREATE INDEX IF NOT EXISTS idx_relationships_source ON memory_relationships(source_memory_id)`,
  `CREATE INDEX IF NOT EXISTS idx_relationships_target ON memory_relationships(target_memory_id)`,
  `CREATE INDEX IF NOT EXISTS idx_relationships_type ON memory_relationships(relationship_type)`,
  `CREATE INDEX IF NOT EXISTS idx_documents_project ON documents(project_id)`,
  `CREATE INDEX IF NOT EXISTS idx_document_chunks_doc ON document_chunks(document_id)`,
];

File: rewrite/crates/db/src/schema.rs
use arrow_schema::{DataType, Field, Schema};
use std::sync::Arc;

/// Schema for the memories table
pub fn memories_schema(vector_dim: usize) -> Arc<Schema> {
  Arc::new(Schema::new(vec![
    Field::new("id", DataType::Utf8, false),
    Field::new("project_id", DataType::Utf8, false),
    Field::new("content", DataType::Utf8, false),
    Field::new("summary", DataType::Utf8, true),
    Field::new("sector", DataType::Utf8, false),
    Field::new("tier", DataType::Utf8, false),
    Field::new("memory_type", DataType::Utf8, true),
    Field::new("importance", DataType::Float32, false),
    Field::new("salience", DataType::Float32, false),
    Field::new("confidence", DataType::Float32, false),
    Field::new("access_count", DataType::UInt32, false),
    Field::new("tags", DataType::Utf8, false),       // JSON array
    Field::new("concepts", DataType::Utf8, false),   // JSON array
    Field::new("files", DataType::Utf8, false),      // JSON array
    Field::new("categories", DataType::Utf8, false), // JSON array
    Field::new("context", DataType::Utf8, true),
    Field::new("session_id", DataType::Utf8, true),
    Field::new("segment_id", DataType::Utf8, true), // Conversation segment ID
    Field::new("scope_path", DataType::Utf8, true), // Code path context
    Field::new("scope_module", DataType::Utf8, true), // Logical module context
    Field::new("created_at", DataType::Int64, false), // Unix timestamp ms
    Field::new("updated_at", DataType::Int64, false),
    Field::new("last_accessed", DataType::Int64, false),
    Field::new("deleted_at", DataType::Int64, true), // Soft delete timestamp
    Field::new("valid_from", DataType::Int64, false),
    Field::new("valid_until", DataType::Int64, true),
    Field::new("is_deleted", DataType::Boolean, false),
    Field::new("content_hash", DataType::Utf8, false),
    Field::new("simhash", DataType::UInt64, false),
    Field::new("superseded_by", DataType::Utf8, true),
    Field::new("decay_rate", DataType::Float32, true), // Cached decay rate
    Field::new("next_decay_at", DataType::Int64, true), // Next scheduled decay
    Field::new("embedding_model_id", DataType::Utf8, true), // Model used for embedding
    Field::new(
      "vector",
      DataType::FixedSizeList(Arc::new(Field::new("item", DataType::Float32, true)), vector_dim as i32),
      true,
    ),
  ]))
}

/// Schema for the code_chunks table
pub fn code_chunks_schema(vector_dim: usize) -> Arc<Schema> {
  Arc::new(Schema::new(vec![
    Field::new("id", DataType::Utf8, false),
    Field::new("project_id", DataType::Utf8, false),
    Field::new("file_path", DataType::Utf8, false),
    Field::new("content", DataType::Utf8, false),
    Field::new("language", DataType::Utf8, false),
    Field::new("chunk_type", DataType::Utf8, false),
    Field::new("symbols", DataType::Utf8, false), // JSON array
    Field::new("start_line", DataType::UInt32, false),
    Field::new("end_line", DataType::UInt32, false),
    Field::new("file_hash", DataType::Utf8, false),
    Field::new("indexed_at", DataType::Int64, false),
    Field::new(
      "vector",
      DataType::FixedSizeList(Arc::new(Field::new("item", DataType::Float32, true)), vector_dim as i32),
      true,
    ),
  ]))
}

/// Schema for the sessions table
pub fn sessions_schema() -> Arc<Schema> {
  Arc::new(Schema::new(vec![
    Field::new("id", DataType::Utf8, false),
    Field::new("project_id", DataType::Utf8, false),
    Field::new("started_at", DataType::Int64, false),
    Field::new("ended_at", DataType::Int64, true),
    Field::new("summary", DataType::Utf8, true),
    Field::new("user_prompt", DataType::Utf8, true),
    Field::new("context", DataType::Utf8, true), // JSON object
  ]))
}

/// Schema for the events table (for event sourcing)
pub fn events_schema() -> Arc<Schema> {
  Arc::new(Schema::new(vec![
    Field::new("id", DataType::Utf8, false),
    Field::new("entity_id", DataType::Utf8, false),
    Field::new("entity_type", DataType::Utf8, false),
    Field::new("event_type", DataType::Utf8, false),
    Field::new("payload", DataType::Utf8, false), // JSON
    Field::new("timestamp", DataType::Int64, false),
  ]))
}

/// Schema for the documents table (ingested docs for search)
pub fn documents_schema(vector_dim: usize) -> Arc<Schema> {
  Arc::new(Schema::new(vec![
    Field::new("id", DataType::Utf8, false),
    Field::new("document_id", DataType::Utf8, false),
    Field::new("project_id", DataType::Utf8, false),
    Field::new("content", DataType::Utf8, false),
    Field::new("title", DataType::Utf8, false),
    Field::new("source", DataType::Utf8, false),
    Field::new("source_type", DataType::Utf8, false),
    Field::new("chunk_index", DataType::UInt32, false),
    Field::new("total_chunks", DataType::UInt32, false),
    Field::new("char_offset", DataType::UInt32, false),
    Field::new("created_at", DataType::Int64, false),
    Field::new("updated_at", DataType::Int64, false),
    Field::new(
      "vector",
      DataType::FixedSizeList(Arc::new(Field::new("item", DataType::Float32, true)), vector_dim as i32),
      true,
    ),
  ]))
}

/// Schema for the session_memories junction table
pub fn session_memories_schema() -> Arc<Schema> {
  Arc::new(Schema::new(vec![
    Field::new("id", DataType::Utf8, false),
    Field::new("session_id", DataType::Utf8, false),
    Field::new("memory_id", DataType::Utf8, false),
    Field::new("usage_type", DataType::Utf8, false), // created, recalled, updated, reinforced
    Field::new("linked_at", DataType::Int64, false), // Unix timestamp ms
  ]))
}

/// Schema for the memory_relationships table
pub fn memory_relationships_schema() -> Arc<Schema> {
  Arc::new(Schema::new(vec![
    Field::new("id", DataType::Utf8, false),
    Field::new("from_memory_id", DataType::Utf8, false),
    Field::new("to_memory_id", DataType::Utf8, false),
    Field::new("relationship_type", DataType::Utf8, false),
    Field::new("confidence", DataType::Float32, false),
    Field::new("valid_from", DataType::Int64, false),
    Field::new("valid_until", DataType::Int64, true),
    Field::new("extracted_by", DataType::Utf8, false),
    Field::new("created_at", DataType::Int64, false),
  ]))
}

/// Schema for the document_metadata table (tracks documents for update detection)
pub fn document_metadata_schema() -> Arc<Schema> {
  Arc::new(Schema::new(vec![
    Field::new("id", DataType::Utf8, false),            // DocumentId
    Field::new("project_id", DataType::Utf8, false),    // Project UUID
    Field::new("title", DataType::Utf8, false),         // Document title
    Field::new("source", DataType::Utf8, false),        // Source path/URL
    Field::new("source_type", DataType::Utf8, false),   // file, url, content
    Field::new("content_hash", DataType::Utf8, false),  // SHA-256 hash for update detection
    Field::new("char_count", DataType::UInt32, false),  // Total character count
    Field::new("chunk_count", DataType::UInt32, false), // Number of chunks created
    Field::new("full_content", DataType::Utf8, true),   // Full document content (optional)
    Field::new("created_at", DataType::Int64, false),   // Unix timestamp ms
    Field::new("updated_at", DataType::Int64, false),   // Unix timestamp ms
  ]))
}

/// Schema for the entities table
pub fn entities_schema() -> Arc<Schema> {
  Arc::new(Schema::new(vec![
    Field::new("id", DataType::Utf8, false),
    Field::new("name", DataType::Utf8, false),
    Field::new("entity_type", DataType::Utf8, false),
    Field::new("summary", DataType::Utf8, true),
    Field::new("aliases", DataType::Utf8, false), // JSON array
    Field::new("first_seen_at", DataType::Int64, false),
    Field::new("last_seen_at", DataType::Int64, false),
    Field::new("mention_count", DataType::UInt32, false),
  ]))
}

/// Schema for the memory_entities junction table
pub fn memory_entities_schema() -> Arc<Schema> {
  Arc::new(Schema::new(vec![
    Field::new("id", DataType::Utf8, false),
    Field::new("memory_id", DataType::Utf8, false),
    Field::new("entity_id", DataType::Utf8, false),
    Field::new("role", DataType::Utf8, false),
    Field::new("confidence", DataType::Float32, false),
    Field::new("extracted_at", DataType::Int64, false),
  ]))
}

/// Schema for the index_checkpoints table (for resuming interrupted indexing)
pub fn index_checkpoints_schema() -> Arc<Schema> {
  Arc::new(Schema::new(vec![
    Field::new("id", DataType::Utf8, false), // Checkpoint ID (project_id + type)
    Field::new("project_id", DataType::Utf8, false), // Project UUID
    Field::new("checkpoint_type", DataType::Utf8, false), // "code" or "document"
    Field::new("processed_files", DataType::Utf8, false), // JSON array of processed file paths
    Field::new("pending_files", DataType::Utf8, false), // JSON array of pending file paths
    Field::new("total_files", DataType::UInt32, false), // Total files to process
    Field::new("processed_count", DataType::UInt32, false), // Files successfully processed
    Field::new("error_count", DataType::UInt32, false), // Files with errors
    Field::new("gitignore_hash", DataType::Utf8, true), // Hash of gitignore rules at start
    Field::new("started_at", DataType::Int64, false), // Unix timestamp ms
    Field::new("updated_at", DataType::Int64, false), // Unix timestamp ms
    Field::new("is_complete", DataType::Boolean, false), // Whether indexing finished
  ]))
}

/// Schema for the segment_accumulators table (tracks work context during extraction)
pub fn segment_accumulators_schema() -> Arc<Schema> {
  Arc::new(Schema::new(vec![
    Field::new("id", DataType::Utf8, false),                    // segment_id
    Field::new("session_id", DataType::Utf8, false),            // Session UUID
    Field::new("project_id", DataType::Utf8, false),            // Project UUID
    Field::new("segment_start", DataType::Int64, false),        // Unix timestamp ms
    Field::new("user_prompts", DataType::Utf8, false),          // JSON array of UserPrompt objects
    Field::new("files_read", DataType::Utf8, false),            // JSON array of paths
    Field::new("files_modified", DataType::Utf8, false),        // JSON array of paths
    Field::new("commands_run", DataType::Utf8, false),          // JSON array of {command, exit_code}
    Field::new("errors_encountered", DataType::Utf8, false),    // JSON array
    Field::new("searches_performed", DataType::Utf8, false),    // JSON array of patterns
    Field::new("completed_tasks", DataType::Utf8, false),       // JSON array of task content
    Field::new("last_assistant_message", DataType::Utf8, true), // Truncated to 10KB
    Field::new("tool_call_count", DataType::UInt32, false),
    Field::new("updated_at", DataType::Int64, false), // Unix timestamp ms
  ]))
}

/// Schema for the extraction_segments table (records extraction runs)
pub fn extraction_segments_schema() -> Arc<Schema> {
  Arc::new(Schema::new(vec![
    Field::new("id", DataType::Utf8, false),                // Segment ID
    Field::new("session_id", DataType::Utf8, false),        // Session UUID
    Field::new("project_id", DataType::Utf8, false),        // Project UUID
    Field::new("trigger", DataType::Utf8, false),           // user_prompt, pre_compact, stop, todo_completion
    Field::new("user_prompts_json", DataType::Utf8, false), // Snapshot of prompts at extraction
    Field::new("files_read_count", DataType::UInt32, false),
    Field::new("files_modified_count", DataType::UInt32, false),
    Field::new("tool_call_count", DataType::UInt32, false),
    Field::new("memories_extracted", DataType::UInt32, false),
    Field::new("extraction_duration_ms", DataType::UInt32, false),
    Field::new("input_tokens", DataType::UInt32, true),
    Field::new("output_tokens", DataType::UInt32, true),
    Field::new("model_used", DataType::Utf8, true),
    Field::new("error", DataType::Utf8, true), // Error message if extraction failed
    Field::new("created_at", DataType::Int64, false), // Unix timestamp ms
  ]))
}

/// Default vector dimensions for embedding models
pub const DEFAULT_VECTOR_DIM: usize = 4096; // qwen3-embedding

#[cfg(test)]
mod tests {
  use super::*;

  #[test]
  fn test_memories_schema() {
    let schema = memories_schema(768);
    assert!(schema.field_with_name("id").is_ok());
    assert!(schema.field_with_name("content").is_ok());
    assert!(schema.field_with_name("vector").is_ok());
  }

  #[test]
  fn test_code_chunks_schema() {
    let schema = code_chunks_schema(768);
    assert!(schema.field_with_name("file_path").is_ok());
    assert!(schema.field_with_name("vector").is_ok());
  }

  #[test]
  fn test_sessions_schema() {
    let schema = sessions_schema();
    assert!(schema.field_with_name("id").is_ok());
    assert!(schema.field_with_name("project_id").is_ok());
  }

  #[test]
  fn test_documents_schema() {
    let schema = documents_schema(768);
    assert!(schema.field_with_name("id").is_ok());
    assert!(schema.field_with_name("content").is_ok());
    assert!(schema.field_with_name("title").is_ok());
    assert!(schema.field_with_name("vector").is_ok());
  }

  #[test]
  fn test_document_metadata_schema() {
    let schema = document_metadata_schema();
    assert!(schema.field_with_name("id").is_ok());
    assert!(schema.field_with_name("source").is_ok());
    assert!(schema.field_with_name("content_hash").is_ok());
  }
}
